### ISSUE TITLE
NodeSeq ++ results in List[Node] in 2.13

### ISSUE THREAD

**Issue by fanf** [scala/scala-xml#392]
I see the following change in behavior between scala 2.12 and 2.13:
**Scala 2.12**:
```
scala> import scala.xml._
import scala.xml._

scala> val a: NodeSeq = <a>Hello</a>
a: scala.xml.NodeSeq = <a>Hello</a>

scala> a ++ <b>hi</b>
res0: scala.xml.NodeSeq = NodeSeq(<a>Hello</a>, <b>hi</b>)
```
**Scala 2.13**:
```
scala> a ++ <b>hi</b>
res0: Seq[scala.xml.Node] = List(<a>Hello</a>, <b>hi</b>)
```

That change has a very deep reach for us, because Liftweb doesn't interpret both type in the same way: it applies template transformation one time for a `NodeSeq` (it's only one template), and N times for `List[Node]` (it's a list of template). Our web application is totally broken, and it's almost impossible to find all places which would need a type ascription (and it would bloat a lot the resulting code.

Is there a way to get back the old behavior? Perhaps we are missing a new implicit? Perhaps the implicit precedence in `NodeSeq` changed and it could be adpated?
Any insight would be appreciated.

**Comment by ashawley** [comment]
Yeah, the code base for 2.13 is new as a result of the changes to the collections library updates, so the code for `NodeSeq` here is also different for 2.12.

If we're going to fix this and make a release, we should make the change to 1.x branch.  It will be easy to cherry-pick to master for a 2.0.0-M2 release.

**Comment by ashawley** [comment]
Here's a first attempt at writing a unit test:

```scala
package scala.xml

import org.junit.Test
import org.junit.Assert.assertEquals
import org.junit.Assert.fail

class NodeSeqTest {

  @Test
  def testAppend: Unit = {
    val a: NodeSeq = <a>Hello</a>
    val b = <b>Hi</b>
    val res: NodeSeq = a ++ b
    assertEquals(NodeSeq.fromSeq(Seq(<a>Hello</a>, <b>Hi</b>)), res)
    a ++ <b>Hi</b> match {
      case _: NodeSeq => assertEquals(2, res.size) // 2.12 and earlier
      case _: List[Node] => fail("Should be NodeSeq") // 2.13
    }
  }
}
```

**Comment by ashawley** [comment]
Since there were a lot of changes to the build during 2.13 milestones, it's difficult to bisect this.  My first pass at it suggests switching to 2.13.0-M4 in 4c7a9454 is the issue.  That raises different questions than I epxected, though.  I'd need to dig further in to this to confirm.

**Comment by jtjeferreira** [comment]
I think NodeSeq needs to override `++` to return `NodeSeq` as explained in https://docs.scala-lang.org/overviews/core/custom-collections.html

> To address this shortcoming, you need to overload the methods that return an IndexedSeq[B] for the case where B is known to be Base, to return an RNA2 instead.

**Comment by fanf** [comment]
I was looking to do so (appendAll and prepend All too, like it's done in
List I believe) but I can't get scala xml project compile, I have tests in
error.
Thanks for the reference doc, I didn't find it before :)

Le mar. 21 janv. 2020 à 21:34, João Ferreira <notifications@github.com> a
écrit :

**Comment by SethTisue** [comment]
>  I can't get scala xml project compile, I have tests in error.

that's a bit disturbing — open a separate ticket on that and let's work through it?

**Comment by ashawley** [comment]
> I think NodeSeq needs to override `++` to return `NodeSeq`

Good catch!  

I've tried to cargo cult that example code in #396.  It seemed to fix the specific issue raised here.

**Comment by fanf** [comment]
@SethTisue: sorry, I wasn't in the correct branch, you can ignore my previous comment.

**Comment by SethTisue** [comment]
@fanf thanks for reporting this! @ashawley has now published 1.3.0 with the fix
